### ISSUE TITLE
[product editor] Applications are shown that are not available

### ISSUE THREAD

**Issue by laeubi** [eclipse-pde/eclipse.pde#1052]
1.   Create a new product
2. Add the eclipse.sdk feature
3. Now in the main tab select the `org.eclipse.equinox.p2.director` application
4. The launch fails with 

> Application "org.eclipse.equinox.p2.director" could not be found in the registry. The applications available are: org.eclipse.e4.ui.workbench.swt.E4Application, org.eclipse.e4.ui.workbench.swt.GenTopic, org.eclipse.help.base.infocenterApplication, org.eclipse.help.base.helpApplication, org.eclipse.help.base.indexTool, org.eclipse.ui.ide.workbench, org.eclipse.equinox.app.error.

The editor should only show these applications that are available.

**Comment by opcoach** [comment]
Yes the list of products and application should be filtered on the available extensions from the dependencies...  Or if we choose one, it should propose to depend on the plugin that provide this application/product.

Could add an 'help' in a information label to explain how to find all the applications and productions available in the target platform (Ctrl Shit A or Cmd Shift A on Mac and then search for org.eclipse.runtime.application|product)

I join a sample project that shows the problem : no dependencies, but all products and applications displayed in the lists in the overview tab of the product editor. 

[test1052.zip](https://github.com/user-attachments/files/17165512/test1052.zip)

**Comment by Cecilia-Rz** [comment]
I'm working on this

**Comment by HannesWell** [comment]
> I'm working on this

Great. You are one of the students from CodeDays, aren't you?

Much success on this task and don't hesitate to ask for help if you need any.

**Comment by KViruz2750** [comment]
I am working on this.

**Comment by KViruz2750** [comment]
> > I'm working on this
> 
> Great. You are one of the students from CodeDays, aren't you?
> 
> Much success on this task and don't hesitate to ask for help if you need any.

I would like some help regarding this problem if that can be given.

**Comment by HannesWell** [comment]
> I would like some help regarding this problem if that can be given.

Sure. Where exactly and what kind of help do you need?
If you struggle to set up your development environment I suggest you have a look at:
https://github.com/eclipse-ide/

**Comment by KViruz2750** [comment]
> > I would like some help regarding this problem if that can be given.
> 
> Sure. Where exactly and what kind of help do you need? If you struggle to set up your development environment I suggest you have a look at: https://github.com/eclipse-ide/

I need help with setting up the program, and just want the steps to put me in the right direction, because I am very confused as to how to approach this problem, and how to actually solve it.

**Comment by vihuynh72** [comment]
I am working on this!

**Comment by vihuynh72** [comment]
These are my teammates: @JZavala210 @krishneetRAJ

My teammates and I are currently working on this issue.

We’re currently investigating an issue where creating a new .product file, selecting the `org.eclipse.equinox.p2.director` application, and launching it results in an error:

Application "org.eclipse.equinox.p2.director" could not be found in the registry.

So, this is the workflow we observed:

- We created a new product configuration.
- In the “features” tab, we noticed eclipse.sdk was not available to select.

<img width="480" height="474" alt="Image" src="https://github.com/user-attachments/assets/57279294-e63c-4798-a694-d9a42d802a0a" />

- We did not manually add Eclipse SDK to the target platform.
- We proceeded to the “launch configuration” and selected org.eclipse.equinox.p2.director as the application.

<img width="719" height="478" alt="Image" src="https://github.com/user-attachments/assets/b35b4da8-998d-40cd-b213-58e47a07f3a0" />

- Launch fails with the above error.

Our assumption is that the missing SDK and/or dependencies are why the selected application cannot be found in the registry.

To investigate deeper, we added debug changes in the `getApplicationNameSet()` method, which controls the dropdown list of available applications in the .product file UI. That function pulls all application extensions using `findExtensions("org.eclipse.core.runtime.applications", true)`. Looking into `findExtensions()`, we noticed that it does not currently check whether the extension is available based on the product's dependencies.

Can you please help us check our understanding? Does our understanding of how the `.product` file and application registration work seem correct? If not, are we missing a step in setting up the right dependencies? We would appreciate any feedback. Thank you very much!

**Comment by HannesWell** [comment]
> The computation of all Plugins included in a launch can be done using (using `osgi=false`):
> 
> [eclipse.pde/ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/BundleLauncherHelper.java](https://github.com/eclipse-pde/eclipse.pde/blob/260640d9827f44452bd0e03d1baea7efe3ab191b/ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/BundleLauncherHelper.java#L110)
> 
> Unfortunately this methods requires a launch configuration as input, so I suggest to create a temporary (hidden) one on the fly to call that method with. To do that you can create a new `org.eclipse.pde.internal.ui.launcher.LaunchAction` for the current product and call `findLaunchConfiguration()`. Maybe we should introduce a new method that always creates a new config and never saves it to prevent it from becoming visible in the UI.

Disregard this part. There is already a method for exactly that purpose:
https://github.com/eclipse-pde/eclipse.pde/blob/260640d9827f44452bd0e03d1baea7efe3ab191b/ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/LaunchAction.java#L98-L104

From the returned set you can get the symbolic-names from all bundles in the launch (and if you want to be very exact, also their IDs).

**Comment by JZavala210** [comment]
I along with @vihuynh72 and @krishneetRAJ have implemented a solution and want to ensure that the solution is one that solves the issue as expected. 

We first got the launched bundles and the symbolic name from all the bundles and then passed the Set of symbolic names directly to the getApplicationNameSet() function directly where we modified the function to take in a Set of strings as seen below. 

<img width="753" height="541" alt="Image" src="https://github.com/user-attachments/assets/48c1255a-ff76-44aa-940c-1343a51bc034" />
 
Of course this created error is other functions like getDefaultApplication() in TargetPlatform.java which uses the original getApplicationNameSet function that doesn't take in a parameter, but that can be solved with an overload of the getApplicationNameSet() function. 

Now for the getApplicationNameSet() function we modified it by adding a if statement to check if the bundle is in the product as seen below. 

<img width="656" height="405" alt="Image" src="https://github.com/user-attachments/assets/eec04c7d-13d3-422e-a885-de0f2146ee1b" />

Now when we run the Plug-in project we made and run it using the Eclipse Run Configuration we made we get a sandbox IDE with the applications list containing: org.eclipse.ui.ide.workbench (which is always added at the end of the  getApplicationNameSet() function). 

Now if you go to the product add a plug-in ant then run it again you will then see the dropdown list has more applications. 

<img width="769" height="164" alt="Image" src="https://github.com/user-attachments/assets/38fd1031-ad96-44d5-91f4-872d8649d9da" />

<img width="769" height="400" alt="Image" src="https://github.com/user-attachments/assets/a00d2e19-8be2-478b-86fe-5f8941ebce08" />

Now in the sandbox IDE if you try to run the product choosing the 'org.eclipse.e4.ui.workbench.swt.E4Application' and run it as an Eclipse Application it fails, but once we added the plug-in 'org.eclipse.e4.tools' in the actual IDE and then ran it using the configuration to create the sandbox IDE and then choose the 'org.eclipse.e4.ui.workbench.swt.E4Application' application and ran it as an Eclipse Application, the launch worked (the application list also includes more applications when adding the 'org.eclipse.e4.tools' plug-in). 

<img width="763" height="440" alt="Image" src="https://github.com/user-attachments/assets/fb1cdd47-7ab7-4734-b387-76016aa33d3c" />

<img width="1114" height="735" alt="Image" src="https://github.com/user-attachments/assets/acb52814-a8e9-4383-838a-90701c6d7454" />

<img width="890" height="545" alt="Image" src="https://github.com/user-attachments/assets/e5d40f35-1ee5-4401-984c-bfd24249d8d6" />

Does our solution fully solve the issue in the way that was expected? We want to certain before we make some final touches and eventually make a pull request. Thank you.
